src/libs/bin_decoder/lv_bin_decoder.c

```diff
@@ -1008,42 +864,149 @@ static lv_result_t decompress_image(lv_image_decoder_dsc_t * dsc, const lv_image
-            lv_cache_unlock();
-            return LV_RESULT_OK;
+        /*The variables should have valid data*/
+        lv_image_dsc_t * image = (lv_image_dsc_t *)dsc->src;
+        if(image->data == NULL) {
+            return false;
+        }
+
+        lv_color_format_t cf = image->header.cf;
+        if(dsc->header.flags & LV_IMAGE_FLAGS_COMPRESSED) {
+            res = decode_compressed(decoder, dsc);
+        }
+        else if(LV_COLOR_FORMAT_IS_INDEXED(cf)) {
+            /*Need decoder data to store converted image*/
+            decoder_data_t * decoder_data = get_decoder_data(dsc);
+            if(decoder_data == NULL) {
+                return false;
+            }
+
+            res = decode_indexed(decoder, dsc);
+        }
+        else if(LV_COLOR_FORMAT_IS_ALPHA_ONLY(cf)) {
+            /*Alpha only image will need decoder data to store pointer to decoded image, to free it when decoder closes*/
+            decoder_data_t * decoder_data = get_decoder_data(dsc);
+            if(decoder_data == NULL) {
+                return false;
+            }
+
+            res = decode_alpha_only(decoder, dsc);
+        }
+        else {
+            /*In case of uncompressed formats the image stored in the ROM/RAM.
+             *So simply give its pointer*/
+
+            decoder_data_t * decoder_data = get_decoder_data(dsc);
+            lv_draw_buf_t * decoded = &decoder_data->c_array;
+            dsc->decoded = decoded;
+            lv_draw_buf_from_image(decoded, image);
+
+            if(decoded->header.stride == 0) {
+                /*Use the auto calculated value from decoder_info callback*/
+                decoded->header.stride = dsc->header.stride;
+            }
+
+            /**
+             * @todo need to convert c-array image stride if not match
+             *
+             * lv_draw_buf_create(); //create new draw buf that meets requirement
+             * lv_draw_buf_copy(); //copy from c-array image to new draw buf
+             */
+            uint32_t stride_expect = lv_draw_buf_width_to_stride(dsc->header.w, dsc->header.cf);
+            if(dsc->header.stride != stride_expect) {
+                LV_LOG_WARN("Stride mismatch");
+#if 0
+                /**
+                 * @fixme ignore for now
+                 */
+                free_decoder_data(dsc);
+                return false;
+#endif
+            }
+
+            res = LV_RESULT_OK;
         }
     }
 
-    lv_cache_unlock();
-    return LV_RESULT_INVALID;
+    if(res != LV_RESULT_OK) {
+        free_decoder_data(dsc);
+        return false;
+    }
+
+    return true;
 }
 
-static void cache_invalidate_cb(lv_cache_entry_t * entry)
+static void bin_decoder_cache_free_cb(lv_image_cache_data_t * cached_data, void * user_data)
 {
+    LV_UNUSED(user_data); /*Unused*/
+
     lv_image_decoder_dsc_t fake = { 0 };
-    fake.user_data = entry->user_data;
+    fake.user_data = cached_data->user_data;
     free_decoder_data(&fake);
```

Comment:
Let separate the `free_decoder_data` with `free_image_cache_entry`. So we can unify it with other decoders.